File: Go/Go_offer/offerII041.go
// Go
// leetcode offerII 041
// https://leetcode.cn/problems/qIsx9U/

type MovingAverage struct {
	Arr []int
	P, Sum, Size int
}


/** Initialize your data structure here. */
func Constructor(size int) MovingAverage {
	return MovingAverage{make([]int, size), 0, 0, 0}
}


func (this *MovingAverage) Next(val int) float64 {
	bigSize := len(this.Arr)

	// 此次之前是否满了
	if this.Size < bigSize {
		// 未满，加值
		this.Size++
		this.Arr[this.P] = val
		this.P++
		this.Sum += val
	} else {
		// 满了，替换值
		this.P %= bigSize
		this.Sum += val - this.Arr[this.P]
		this.Arr[this.P] = val
		this.P++
	}
	
	return float64(this.Sum) / float64(this.Size)
}


/**
 * Your MovingAverage object will be instantiated and called as such:
 * obj := Constructor(size);
 * param_1 := obj.Next(val);
 */

File: cpp/cpp_offer/offerII041.cpp
// C++
// leetcode offerII 041
// https://leetcode.cn/problems/qIsx9U/

class MovingAverage {
public:
    /** Initialize your data structure here. */
    int size;
    vector<int> arr;
    int sum, p;
    MovingAverage(int size_) {
    	sum = 0;
    	size = 0;
    	arr.resize(size_);
    	p = 0;
    }
    
    double next(int val) {
    	if (size < arr.size()) {
    		size++;
    		sum += val;
    		arr[p] = val;
    		p = (p + 1) % arr.size();
    		double d_sum = (double)sum;
    		return d_sum / size;
    	} else {
    		sum += val - arr[p];
    		arr[p] = val;
    		p = (p + 1) % arr.size();
    		double d_sum = (double)sum;
    		return d_sum / size;
    	}
    }
};

/**
 * Your MovingAverage object will be instantiated and called as such:
 * MovingAverage* obj = new MovingAverage(size);
 * double param_1 = obj->next(val);
 */